docs/builder/ref.md

```diff
@@ -0,0 +1,290 @@
+# Builder Library Reference
+
+## Top-Level Program Structure
+
+Here's the general structure of a program that uses the builder to generate Calyx code.
+
+```python
+# import the builder library
+import calyx.builder as cb
+
+
+# define my_subcomponent
+def add_my_subcomponent(prog):
+    # subcomponent definition here
+
+
+# define my_component
+def my_component(prog, my_subcomponent): 
+    # add the component to the program
+    my_component = prog.component("my_component")
+
+    # add my_subcomponent as a cell of my_component
+    my_subcomponent = my_component.cell("my_subcomponent", my_subcomponent)
+
+    # define a my_component group
+    with my_component.group("my_group") as my_group:
+      # group assignments here 
+
+  my_component.control += [my_group]
+
+
+# assemble the program
+def build():
+    prog = cb.Builder()
+    my_subcomponent = add_my_subcomponent(prog)
+    add_main(prog, my_subcomponent)
+
+    # return the generated program
+    return prog.program
+
+
+# emit the program
+if __name__ == "__main__":
+    build().emit()
+```
+
+## Components
+
+### Defining Components
+
+You can define a component by calling the `Builder().component()` method.
+
+```python
+prog = cb.Builder()
+prog.component("my_component")
+```
+
+Components can also be instantiated with a list of cells.
+
+```python
+
+ adders = [adder_1, adder_2, adder_3, adder_4] = [
+        my_component.add(name, 32) for name in ["1", "2", "3", "4"]
+    ]
+my_component = prog.component("my_component", cells=adders)
+```
+
+### Retrieving Components
+
+To reference a component you did not [store a handle][hndl], you can do so later with the `Builder().get_component()` method.
+
+```python
+prog = cb.Builder()
+prog.component("my_component")
+# a few lines later 
+my_component = prog.get_component("my_component")
+```
+
+### Defining component inputs and outputs
+
+Components can be given input and output ports. All you have to do is specify the name of the port and its size.
+
+```python
+my_component.input("my_input", 32)
+my_component.output("my_output", 32)
+```
+
+You can access the input and output ports of a component within the definition of a component using the standard `this.port` syntax.
+
+```python
+def add_my_component(prog):
+    my_component = prog.component("my_component")
+
+    this = my_component.this()
+    with my_component.group("my_group")
+      # some other assignments
+        this.out = sum.out
+```
```

Comment:
you're right, good catch! will amend

Reply:
seconding @anshumanmohan 

I'd show the declaration of the `output(...)` fn in this example and then just assign to it's return val in the assignment, since that's the recommended approach.